### PR TITLE
Basic Functionality of the tvp  Debugger

### PR DESCRIPTION
**Pull request by trickarcher**

### CODE REVIEW

File: src/debugger/CMakeLists.txt [hunk 1]
```diff
@@ -2,11 +2,14 @@ cmake_minimum_required(VERSION 3.5.1)
...
+  )
 
 add_library(${PROJECT_NAME} STATIC ${SOURCE_FILES})
 
-include_directories(${MODULE_INCLUDE_DIRS})
+include_directories(${MODULE_INCLUDE_DIRS} include)
 target_include_directories(${PROJECT_NAME} PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
 )
+target_link_libraries(${PROJECT_NAME} cxxopts)
```
Comment (venkat24) [inline src/debugger/CMakeLists.txt:15]: Why is this necessary?

File: src/debugger/include/debugger/debugger_interface.h [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @file debugger_interface.h
+ * Declares the DebuggerCore Interface Class for CLI Operations
+ */
+
+#include "debugger/debugger.fwd.h"
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger_interface.h]: This include is not required

File: src/debugger/src/cli_debugger.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	if (str.compare("h") == 0 || str.compare("help") == 0) {
+		print_help();
+		return true;
+	}
+	if (str.substr(0, 1).compare("p") == 0) {
+		auto code_string = debugger_core->peek(strtoul(str.substr(2).c_str(), nullptr, 10));
+		cout << code_string << "\n";
+		return true;
+	}
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: This parsing logic seems a bit tacky, especially with the string splitting.

Can we use cxxopts once again for this?

File: src/debugger/src/debugger.cpp [hunk 4]
```diff
@@ -1,27 +1,158 @@
...
+	// A safety measure to ensure breakpoints are not repeated
+	auto found_iterator =
+	    std::find(breakpoints.begin(), breakpoints.end(), breakpoint);
+	if (found_iterator == breakpoints.end()) {
+		breakpoints.push_back(breakpoint);
+		return 1;
+	} else {
+		return 0;
+	}
+	return 2;
```
Comment (venkat24) [inline src/debugger/src/debugger.cpp]: This is unreachable code.

Also, these set and remove methods don't need to return status codes. Perhaps a success boolean could be useful, but we don't need to return an int. You can just simplify it to 

```cpp
    if (std::find(breakpoints.begin(), breakpoints.end(), breakpoint) == breakpoints.end()) {
        breakpoints.push_back(breakpoint);
    }
```

Also, these breakpoint containers can be `std::unordered_set` instead, so that this lookup operation is O(1).

File: src/debugger/include/debugger/debugger.h [hunk 5]
```diff
@@ -29,47 +30,106 @@ namespace debugger {
...
+
+	/**
+	 * A vector container for storing breakpoints at GameBoy cycles
+	 */
+	std::vector<ClockCycles> tick_breakpoints;
+
+	/**
+	 * A vector container iterator to keep track of breakpoints
+	 */
+	std::vector<Address>::iterator breakpoints_iter = breakpoints.begin();
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: Is this used?

File: src/debugger/include/debugger/debugger.h [hunk 6]
```diff
@@ -29,47 +30,106 @@ namespace debugger {
...
+
+	/**
+	 * @brief Adds breakpoints to the Vector Container, skips if breakpoint
+	 * already present
+	 * @param breakpoint of Instructions, ticks and clocks
+	 *
+	 * returns 0 if Breakpoint was already present, so not added;
+	 * returns 1 if Breakpoint was successfully inserted;
+	 * returns 2 default return value. Will not reach here;
+	 * I chose uint8_t as a return type so that we can add to the error codes
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: My bad, I'd already added that comment above, but again, why do we need error codes here?

Even if a breakpoint was already present, can't we still just ignore it and return success? At the moment, even a boolean error code doesn't seem necessary.

File: src/debugger/include/debugger/debugger.h [hunk 7]
```diff
@@ -29,47 +30,106 @@ namespace debugger {
...
+
+	// Get current debugger state
+	virtual vector<Address> get_breakpoints();
+	virtual vector<ClockCycles> get_cycle_breakpoints();
+	virtual vector<ClockCycles> get_tick_breakpoints();
+
+	virtual void run();
+	virtual void step();
+
+	virtual std::string peek(uint32_t lines);
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: Perhaps return a std::vector of strings? The debugger core shouldn't do any formatting. That should be done by the UI layer.

You can move the newline printing logic to the CliDebugger class instead.

File: src/debugger/include/debugger/debugger.h [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
- * Declares the Debugger Class
+ * Declares the DebuggerCore Class
  */
 
 #pragma once
 
 #include "cartridge/cartridge.h"
 #include "cpu/cpu.h"
 #include "cpu/utils.h"
+#include "debugger/cli_debugger.h"
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: Shouldn't be necessary

File: src/debugger/include/debugger/debugger.h [hunk 9]
```diff
@@ -1,13 +1,14 @@
 /**
  * @file debugger.h
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: Rename the file as well

File: src/debugger/include/debugger/debugger.h [hunk 10]
```diff
@@ -29,47 +30,106 @@ namespace debugger {
...
+
+	/**
+	 * A vector container iterator to keep track of breakpoints
+	 */
+	std::vector<Address>::iterator breakpoints_iter = breakpoints.begin();
+
+	/**
+	 * A flag to check if we have reached the breakpoint
+	 */
+	bool is_breakpoint = false;
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: Is this used?

File: src/debugger/src/cli_debugger.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    : debugger_core(debugger_core) {}
+
+void CliDebugger::tick() {
+	cout << "Enter here: ";
+	getline(std::cin, str);
+	bool status = parse_command(str);
+	if (!status) {
+		cout << "Not a valid Command, press h for help" << endl;
+		tick();
+	}
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: The name `parse_command` here is a little bit misleading, since it is not just parsing the command, but also executing the actions. `run_command` might be a better name.

Also, this could be readable as an iterative `do..while` loop than as a recursive call. A finer detail is also that since this is a potentially unbounded recursion, if the user kept inputting wrong commands, you'd eventually overflow the call stack :)

File: src/debugger/src/debugger.cpp [hunk 12]
```diff
@@ -1,27 +1,158 @@
...
+		 */
+		tick();
+	}
+
+	/**
+	 * Return to the Interface if no breakpoints are there to process
+	 */
+	if (breakpoints.empty() && tick_breakpoints.empty() &&
+	    cycle_breakpoints.empty()) {
+		return;
```
Comment (venkat24) [inline src/debugger/src/debugger.cpp]: I think the behavior here should be, if there are no breakpoints, just keep running. This shouldn't be a special case, right?

File: src/debugger/src/debugger.cpp [hunk 13]
```diff
@@ -1,27 +1,158 @@
...
+	 * breakpoints and increment the counter by one, once a breakpoint is
+	 * reached. If all the breakpoints are processed, we wait for another
+	 * breakpoint and keep the iterator in it's position. This will work for
+	 * System tick breakpoints and CPU cycles, I have my doubts about
+	 * Instruction breakpoints so I am going to use a STL search function for
+	 * instruction breakpoints;
+	 */
+	while (std::find(breakpoints.begin(), breakpoints.end(),
+	                 gameboy->cpu->pc->get()) == breakpoints.end() ||
+	       std::find(processed_breakpoints.begin(), processed_breakpoints.end(),
```
Comment (venkat24) [inline src/debugger/src/debugger.cpp]: Why is `processed_breakpoints` needed? A breakpoint should not be removed when it is hit - the same instruction breakpoint can be called multiple times inside a loop, and we often want to pause each time it is called and hit `run` to loop back again.

File: src/main/main.cpp [hunk 14]
```diff
@@ -59,15 +59,17 @@ int main(int argc, char *argv[]) {
...
-		// Start Gameboy normally
+		// Start GameBoy normally
 		for (auto i = 0; /*Infinite Loop*/; i++) {
 			gameboy->tick();
 		}
 	} else {
-		// Start Gameboy with Debugger
-		auto debugger = std::make_unique<Debugger>(std::move(gameboy));
+		// Start GameBoy with DebuggerCore
+		auto debugger = std::make_shared<DebuggerCore>(std::move(gameboy));
```
Comment (venkat24) [inline src/main/main.cpp]:
```suggestion
		auto debugger_core = std::make_shared<DebuggerCore>(std::move(gameboy));
```

File: src/debugger/include/debugger/cli_debugger.h [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+#include "debugger.h"
+#include "debugger/debugger.fwd.h"
+#include "debugger/debugger_interface.h"
+
+namespace debugger {
+
+class CliDebugger : public IDebugger {
+  private:
+	std::shared_ptr<DebuggerCore> debugger_core;
```
Comment (venkat24) [inline src/debugger/include/debugger/cli_debugger.h]: Does this need to be a `shared_ptr`? Since this is the sole owner of the ```DebuggerCore``` object, this can be made a ```unique_ptr```.

File: src/debugger/include/debugger/cli_debugger.h [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+	std::string str;
+	std::string command_type;
+	std::string command;
+	std::string command_value;
```
Comment (venkat24) [inline src/debugger/include/debugger/cli_debugger.h]: These can be local variables inside the `run_command` method

File: src/debugger/include/debugger/debugger.h [hunk 17]
```diff
@@ -29,47 +30,106 @@ namespace debugger {
...
+	virtual vector<Address> get_breakpoints();
+	virtual vector<ClockCycles> get_cycle_breakpoints();
+	virtual vector<ClockCycles> get_tick_breakpoints();
+
+	virtual void run();
+	virtual void step();
+
+	virtual std::string peek(uint32_t lines);
+
+	friend class IDebugger;
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger.h]: This should not be necessary.

The interface IDebugger can never be instantiated, and the CliDebugger class shouldn't be `friend`s with DebuggerCore.

File: src/debugger/include/debugger/debugger_interface.h [hunk 18]
```diff
@@ -0,0 +1,18 @@
+/**
+ * @file debugger_interface.h
+ * Declares the DebuggerCore Interface Class for CLI Operations
```
Comment (venkat24) [inline src/debugger/include/debugger/debugger_interface.h]:
```suggestion
 * Declares the Debugger Interface Class
```

This is not exclusively for CLI operations, no?

File: src/memory/src/memory.cpp [hunk 19]
```diff
@@ -67,7 +67,8 @@ uint8_t Memory::read(Address address) const {
 	// Sound Controller Registers
 	if (address_in_range(address, 0xFF26, 0xFF10)) {
 		// TODO: Sound Controller
-		Log::warn("Attempt to read from sound register " + num_to_hex(address));
+		//		Log::warn("Attempt to read from sound register " +
+		// num_to_hex(address));
```
Comment (venkat24) [inline src/memory/src/memory.cpp]: These commented out blocks can be removed

**Review by venkat24 — CHANGES_REQUESTED** [review]

File: src/debugger/CMakeLists.txt (hunk 1 above)
Comment (trickarcher) [reply src/debugger/CMakeLists.txt:15]: I was trying something with cxxopts, forgot to remove it from CMakeLists.txt

File: src/debugger/include/debugger/debugger_interface.h (hunk 2 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger_interface.h]: I thought of making IDebugger a friend class of DebuggerCore. But then I thought it's not necessary because we have public helper functions to retrieve private members if required.

Comment (venkat24) [reply src/debugger/include/debugger/debugger_interface.h]: Ah okay.

An interface is a class you're never going to instantiate, so a friend class for it doesn't do much.

File: src/debugger/src/cli_debugger.cpp (hunk 3 above)
Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp]: Can we? I went through cxxopts lib but it seems that you can give only command-line arguments during the invocation of the executable, {argc, argv}. Personally, I too am not a fan of the printing logic, I am trying to find some other way to tackle this.

File: src/debugger/include/debugger/debugger_interface.h (hunk 2 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger_interface.h]: Ah, okay got it.

File: src/debugger/src/cli_debugger.cpp (hunk 3 above)
Comment (venkat24) [reply src/debugger/src/cli_debugger.cpp]: We can construct the argc/argv manually from the input command string.

Comment (venkat24) [reply src/debugger/src/cli_debugger.cpp]: You can make a helper function that does something similar to this 

https://stackoverflow.com/a/1511885

Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp]: Ah shoot. Yeah we can do that for sure. I'll try this out now.

Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp]: I replaced the printing logic with cxxopts, it looks pretty robust now although I need to add try catch blocks for exceptions thrown by cxxopts.

File: src/debugger/src/debugger.cpp (hunk 4 above)
Comment (trickarcher) [reply src/debugger/src/debugger.cpp]: Thanks for this; I had never heard about unordered_set before. Looks like I have some reading up to do on the implementation, it seems pretty heavy.

File: src/debugger/include/debugger/debugger.h (hunk 5 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: Yeah, no. I'll depend on the find operation from now on since the complexity is O(1), it would be an unnecessary space complexity wastage if I kept 3 iterators for the breakpoints.

File: src/debugger/include/debugger/debugger.h (hunk 6 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: I guess, but I feel it would be better to know what's going on inside. It is kinda unnecessary but it's just doesn't feel right to return success if something's repeated and not added.

File: src/debugger/include/debugger/debugger.h (hunk 7 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: Agreed.

Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: I rather use `map` because having a instruction address corresponding to the instruction mnemonic, it helps a lot while setting breakpoints. What do you think?

File: src/debugger/include/debugger/debugger.h (hunk 8 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: Yup, included the header in main.cpp instead

File: src/debugger/include/debugger/debugger.h (hunk 9 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: Done.

File: src/debugger/src/cli_debugger.cpp (hunk 11 above)
Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp]: Yeah, you're right. I guess it slipped my mind. Sorry!

File: src/debugger/include/debugger/debugger.h (hunk 7 above)
Comment (venkat24) [reply src/debugger/include/debugger/debugger.h]: Yup, that's cool

File: src/debugger/include/debugger/debugger.h (hunk 10 above)
Comment (anish0x2a) [reply src/debugger/include/debugger/debugger.h]: Yup, I use this flag to avoid getting stuck at the sam breakpoint.

File: src/debugger/src/debugger.cpp (hunk 12 above)
Comment (trickarcher) [reply src/debugger/src/debugger.cpp]: Yeah, removed it.

File: src/debugger/src/debugger.cpp (hunk 13 above)
Comment (trickarcher) [reply src/debugger/src/debugger.cpp]: Yeah, I see the flaw in my logic.

File: src/debugger/include/debugger/cli_debugger.h (hunk 16 above)
Comment (trickarcher) [reply src/debugger/include/debugger/cli_debugger.h]: Yeah. I don't need those variables any more.

File: src/debugger/include/debugger/debugger.h (hunk 17 above)
Comment (trickarcher) [reply src/debugger/include/debugger/debugger.h]: Yeah Okay, Got it.

File: src/debugger/src/cli_debugger.cpp [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+	    ("run","Run the ROM till it encounters a breakpoint")
+	    ("step", "Execute one System tick")
+	    ("bp-set", "Set a breakpoint",cxxopts::value<string>()->default_value("NULL"))
+	    ("bp-remove", "Remove a breakpoint",cxxopts::value<string>()->default_value("NULL"))
+	    ("bp_cycles-set", "Set a CPU cycle breakpoint",cxxopts::value<ClockCycles>()->default_value("0"))
+	    ("bp_cycles-remove", "Remove a breakpoint",cxxopts::value<ClockCycles>()->default_value("0"))
+	    ("bp_ticks-set", "Set a breakpoint",cxxopts::value<ClockCycles>()->default_value("0"))
+	    ("bp_ticks-remove", "Remove a breakpoint",cxxopts::value<ClockCycles>()->default_value("0"))
+	    ("peek", "Peek into the next \'x\' lines of the ROM",cxxopts::value<uint32_t>()->default_value("0"))
+	    ("bp-view","View All Instruction Breakpoints")
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: nit: single space after comma

File: src/debugger/src/cli_debugger.cpp [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+	    ("bp_ticks-view", "View all Tick Breakpoints")
+	    ("bp_cycles-view", "View all CPU Cycle Breakpoints")
+        ("help", "Print this information");
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp:25]: nit: indentation

File: src/debugger/src/cli_debugger.cpp [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+
+CliDebugger::CliDebugger(std::unique_ptr<DebuggerCore> debugger_core)
+    : debugger_core(move(debugger_core)),
+      command_parser("tdb", "The tvp Debugger") {
+
+	// clang-format off
+	command_parser.add_options()
+	    ("run","Run the ROM till it encounters a breakpoint")
+	    ("step", "Execute one System tick")
+	    ("bp-set", "Set a breakpoint",cxxopts::value<string>()->default_value("NULL"))
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: An empty string is perhaps a better choice for a default value, it's best to avoid `NULL`

File: src/debugger/src/cli_debugger.cpp [hunk 23]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+void CliDebugger::tick() {
+	std::string str;
+	do {
+		cout << "(tdb) ";
+		getline(std::cin, str);
+
+		/**
+		 * A hack to make CLI parsing compatible with cxxopts.
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: nit: Short, in-code comments should preferably be single line comments with `//`

File: src/debugger/src/cli_debugger.cpp (hunk 22 above)
Comment (venkat24) [reply src/debugger/src/cli_debugger.cpp]: It lets you use `.empty()` as well

File: src/debugger/src/cli_debugger.cpp [hunk 24]
```diff
@@ -0,0 +1,154 @@
...
+			cout << i << " ";
+		}
+		cout << "\n";
+	}
+
+	return true;
+}
+
+void CliDebugger::cmd_attributes_gen(std::string command) {
+	argv_vec.clear();
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: These members should be local variables of this method.

File: src/debugger/src/debugger_core.cpp [hunk 25]
```diff
@@ -0,0 +1,121 @@
...
+
+void DebuggerCore::run() {
+	if (is_breaking) {
+		is_breaking = false;
+		tick();
+	}
+
+	while (breakpoints.find(gameboy->cpu->pc->get()) == breakpoints.end() &&
+	       tick_breakpoints.find(ticks) == tick_breakpoints.end() &&
+	       cycle_breakpoints.find(gameboy->gpu->current_cycles) ==
```
Comment (venkat24) [inline src/debugger/src/debugger_core.cpp]: Why gpu?

File: src/debugger/src/debugger_core.cpp [hunk 26]
```diff
@@ -0,0 +1,121 @@
...
+	           cycle_breakpoints.end()) {
+		tick();
+	}
```
Comment (venkat24) [inline src/debugger/src/debugger_core.cpp]: I think you can clean up this conditional here, and also invert it if that makes more logical sense. You can pull out each breakpoint condition like

```cpp
    auto is_breakpoint_hit = breakpoints.find(gameboy->cpu->pc->get()) != breakpoints.end();
```

File: src/debugger/src/debugger_core.cpp [hunk 27]
```diff
@@ -0,0 +1,121 @@
...
+
+void DebuggerCore::step() { tick(); }
+
+std::map<Address, std::string> DebuggerCore::peek(uint32_t lines) {
+	auto pc_contents = gameboy->cpu->pc->get();
+	std::map<Address, std::string> code_map;
+	for (auto i = 0; i < lines; i++) {
+		code_map.insert(
+		    {+pc_contents, get_mnemonic(gameboy->memory->read(pc_contents))});
+		pc_contents++;
```
Comment (venkat24) [inline src/debugger/src/debugger_core.cpp]: This implementation is incorrect, because not every instruction occupies only 1 byte. Hence, we also need information about the length of each instruction before we can implement peek.

For the time being, you can just leave the peek method unimplemented and throw and error, and address this in a later PR.

File: src/memory/src/memory.cpp (hunk 19 above)
Comment (venkat24) [reply src/memory/src/memory.cpp]: Since the logs are now disabled, you can leave the warnings in place.

You can reset these changes in memory.cpp

File: src/util/src/helpers.cpp [hunk 28]
```diff
@@ -27,6 +27,8 @@ template string num_to_hex(int8_t i);
 template string num_to_hex(int16_t i);
 template string num_to_hex(int32_t i);
 
+Address string_to_hex(string num) { return strtoul(num.c_str(), nullptr, 16); }
```
Comment (venkat24) [inline src/util/src/helpers.cpp]: `string_to_address` would be a better name for this method.

File: src/debugger/src/cli_debugger.cpp [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+		for (auto i : cycles_bp) {
+			cout << i << " ";
+		}
+		cout << "\n";
+	}
+
+	return true;
+}
+
+void CliDebugger::cmd_attributes_gen(std::string command) {
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp]: This can return a tuple of `(int argc, char **argv)`

With C++17 structured bindings, you can unpack them and declare on the caller side in a single line too.

```cpp
auto [argc, argv] = cmd_attributes_gen(command);
```
Handy!

Also, perhaps like `generate_argv_from_string` is a better name?

**Review by venkat24 — CHANGES_REQUESTED** [review]

File: src/debugger/src/cli_debugger.cpp (hunk 29 above)
Comment (venkat24) [reply src/debugger/src/cli_debugger.cpp]: `cmd_string_to_argv` would as well, your preference.

File: src/debugger/src/cli_debugger.cpp [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+			cout << hex << i << " ";
+		}
+		cout << "\n";
+		return true;
+	}
+	if (parsed_args["bp_ticks-view"].as<bool>()) {
+		auto tick_bp = debugger_core->get_tick_breakpoints();
+		cout << "The tick breakpoints are: ";
+		for (auto i : tick_bp) {
+			cout << i << " ";
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp:69]: For anywhere you're printing hex numbers, you can use `num_to_hex`, which is a helper that returns a string representation like `0x00F2`

File: src/debugger/src/cli_debugger.cpp [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+
+	do {
+		cout << "(tdb) ";
+		getline(std::cin, str);
+
+		//  A hack to make CLI parsing compatible with cxxopts.
+		str = "./tdb --" + str;
+		auto [argc, argv] = cmd_string_to_argv(str);
+
+		status = run_command(argc, argv);
```
Comment (venkat24) [inline src/debugger/src/cli_debugger.cpp:41]: This can be a local variable.

As a rule - If a variable is not required outside the context of just one function, like `status` here, it can be a local variable instead of a class member.

**Review by venkat24 — CHANGES_REQUESTED** [review]

File: src/debugger/src/cli_debugger.cpp (hunk 30 above)
Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp:69]: Sure, thanks for the heads up!

File: src/debugger/src/cli_debugger.cpp (hunk 31 above)
Comment (trickarcher) [reply src/debugger/src/cli_debugger.cpp:41]: Understood. I will keep this in mind for future PR's.

**Comment by venkat24** [comment]
Can you please run clang-format?

Everything else looks good!

**Comment by trickarcher** [comment]
Yup, done!

**Review by venkat24 — APPROVED** [review]

**Comment by venkat24** [comment]
It's super cool that we have a debugger now, this is amazing work @anish0x2a !
